### PR TITLE
refactor(goal_planner): refactoring plan flow and add post process

### PR DESCRIPTION
**Pull request by kosuke55**
## Description

refactoring plan flow

```
  /*　
   * state transitions and plan functions used in each state
   *
   * +--------------------------+
   * | RUNNING                  |
   * | plnaPullOverAsCandidate()|
   * +------------+-------------+
   *              | hasDecidedPath()
   *  2           v
   * +--------------------------+
   * | WAITING_APPROVAL         |
   * | planPullOverAsCandidate()|
   * +------------+-------------+
   *              | isActivated()
   *  3           v
   * +--------------------------+
   * | RUNNING                  |
   * | planPullOverAsOutput()   |
   * +--------------------------+
   */
```

## Related links

before this PR merge 
- https://github.com/autowarefoundation/autoware.universe/pull/5546

## Tests performed

psim

evaluator_description: refactor/goal_planner_plan
2023/11/10 https://evaluation.tier4.jp/evaluation/reports/4030409f-6c74-543f-b949-30d8fcaec093/?project_id=prd_jt

evaluator_description: refactor/goal_planner_plan
2023/11/13 https://evaluation.tier4.jp/evaluation/reports/89a8ffc6-8b02-52b8-a77d-3f66c1fc04be/?project_id=prd_jt

evaluator_description: refactor/goal_planner_plan
2023/11/14 https://evaluation.tier4.jp/evaluation/reports/32b6bef1-9217-5383-ab16-0f84297be6c5/?project_id=prd_jt

## Interface changes

none

## Effects on system behavior

none

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by kyoichi-sugahara** [comment]
@kosuke55 
can you add TODO(someone) for writing flow of the module in order not to forget?
> Running
planPullOverAsCandidate()
↓hasDecidedePath()
WAITING_APPROVAL
planPullOverAsCandidate()
↓ isActivated()
Running
planPullOverAsOutput()

File: planning/behavior_path_planner/src/scene_module/goal_planner/goal_planner_module.cpp [hunk 1]
```diff
@@ -253,12 +253,27 @@ void GoalPlannerModule::updateData()
...
+  resetPathReference();
+  path_reference_ = getPreviousModuleOutput().reference_path;
 
-  // set current road lanes, pull over lanes, and drivable lane
-  setLanes();
+  updateOccupancyGrid();
 
   generateGoalCandidates();
+
+  if (!isActivated()) {
```
Comment (kyoichi-sugahara) [inline planning/behavior_path_planner/src/scene_module/goal_planner/goal_planner_module.cpp:264]: @kosuke55
could you add comment why early return when the module is activated?

**Comment by kyoichi-sugahara** [comment]
@kosuke55 
could you resolve conflicts?

**Comment by kyoichi-sugahara** [comment]
thanks!!!!

File: planning/behavior_path_planner/src/scene_module/goal_planner/goal_planner_module.cpp [hunk 2]
```diff
@@ -888,22 +879,50 @@ void GoalPlannerModule::decideVelocity()
...
+
+  return output;
+}
+
+BehaviorModuleOutput GoalPlannerModule::planPullOverAsOutput()
+{
+  // if pull over path candidates generation is not finished, use previous module output
+  if (thread_safe_data_.get_pull_over_path_candidates().empty()) {
+    return getPreviousModuleOutput();
+  }
```
Comment (kyoichi-sugahara) [inline planning/behavior_path_planner/src/scene_module/goal_planner/goal_planner_module.cpp:926]: this check can be moved in the beginning of `planPullOver()` function or somewhere?
this is implemented in the same way in `planPullOverAsCandidate()`

**Comment by kosuke55** [comment]
added flow in https://github.com/autowarefoundation/autoware.universe/pull/5554/commits/20259e296524a51378abc37a6cec94b2a6528be4

**Review by kyoichi-sugahara — APPROVED** [review]
LGTM!

File: planning/behavior_path_planner/src/scene_module/goal_planner/goal_planner_module.cpp (hunk 1 above)
Comment (kosuke55) [reply planning/behavior_path_planner/src/scene_module/goal_planner/goal_planner_module.cpp:267]: fixed in https://github.com/autowarefoundation/autoware.universe/pull/5554/commits/36d4457d3318ed9207c8f94a6c8c21be6e5328ff#diff-13561094b202845230024ce2547446a6d04d30789dda6e9dbe8b08671149ee96R267

File: planning/behavior_path_planner/src/scene_module/goal_planner/goal_planner_module.cpp (hunk 2 above)
Comment (kosuke55) [reply planning/behavior_path_planner/src/scene_module/goal_planner/goal_planner_module.cpp:929]: this is needed here for the case from `planPullOverAsCandidate` is called `planWaitingApproval`

**Pull request merged**